### ISSUE TITLE
TessGate::render() calls produce no output after Tess::drop() has been called

### ISSUE THREAD

**Issue by austinjones** [phaazon/luminance-rs#212]
Hello,
I love the library!  I've been looking for a library that provides more control over fragment shaders than piston2d-graphics, but doesn't take a thousand lines of code like gfx-hal does. 
 It's also very cool that you can write code which is generic over the ColorSlot, which allows code reuse across 'preview' and 'render to file' methods.  I'm using it in https://github.com/austinjones/dali-rs for generative art.  

I found a killer bug that took me a day to track down.

If you issue multiple render calls with Tess instances that have short lifetimes, only the first one actually renders.  Any calls to TessGate::render() after Tess::drop() is invoked run, but produce no output.

I'm using luminance 0.33, luminance-derive 0.2.0, and luminance-glfw 0.7.0 on OSX.

Here is a program to reproduce.  You can see where the second triangle should be by changing the bug_option to BugOption::Working.

```use core::borrow::BorrowMut;
use std::path::Path;

use image::GenericImageView;
use luminance::blending::Equation::Additive;
use luminance::blending::Factor::{SrcAlpha, SrcAlphaComplement};
use luminance::context::GraphicsContext;
use luminance::depth_test::DepthTest;
use luminance::framebuffer::Framebuffer;
use luminance::pixel::{R32F, RGBA32F};
use luminance::render_state::RenderState;
use luminance::shader::program::Program;
use luminance::tess::{Mode, TessBuilder};
use luminance::texture::{Dim2, Flat, GenMipmaps, Sampler, Texture};
use luminance_derive::{Semantics, Vertex};
use luminance_glfw::{Action, GlfwSurface, Key, Surface, WindowDim, WindowEvent, WindowOpt};

const SIMPLE_VS: &'static str = "
in vec2 co;
in vec3 color;

out vec3 v_color;

void main() {
  gl_Position = vec4(co, 0., 1.);
  v_color = color;
}";


const SIMPLE_FS: &'static str = "
in vec3 v_color;

out vec4 frag;

void main() {
  frag = vec4(v_color, 1.);
}";

#[derive(Clone, Copy, Debug, Eq, PartialEq, Semantics)]
pub enum Semantics {
    // reference vertex positions with the co variable in vertex shaders
    #[sem(name = "co", repr = "[f32; 2]", wrapper = "VertexPosition")]
    Position,
    // reference vertex colors with the color variable in vertex shaders
    #[sem(name = "color", repr = "[f32; 3]", wrapper = "VertexColor")]
    Color,
}

#[repr(C)]
#[derive(Clone, Copy, Debug, PartialEq, Vertex)]
#[vertex(sem = "Semantics")]
pub struct Vertex {
    pub pos: VertexPosition,
    pub rgb: VertexColor,
}

// Only one triangle this time.
const TRI_1_VERTICES: [Vertex; 3] = [
    Vertex { pos: VertexPosition::new([0.0, -0.5]), rgb: VertexColor::new([0.0, 1.0, 0.0]) },
    Vertex { pos: VertexPosition::new([-0.5, 0.5]), rgb: VertexColor::new([0.0, 1.0, 0.0]) },
    Vertex { pos: VertexPosition::new([-1.0, -0.5]), rgb: VertexColor::new([0.0, 1.0, 0.0]) },
];

const TRI_2_VERTICES: [Vertex; 3] = [
    Vertex { pos: VertexPosition::new([1.0, -0.5]), rgb: VertexColor::new([1.0, 0.0, 0.0]) },
    Vertex { pos: VertexPosition::new([0.5, 0.5]), rgb: VertexColor::new([1.0, 0.0, 0.0]) },
    Vertex { pos: VertexPosition::new([0.0, -0.5]), rgb: VertexColor::new([1.0, 0.0, 0.0]) },
];

enum BugOption {
    Working,
    Broken,
}

fn main() {
    let bug_option = BugOption::Broken;

    let render_size = [800, 500];
    let mut surface = GlfwSurface::new(
        WindowDim::Windowed(render_size[0], render_size[1]),
        "Hello, world!",
        WindowOpt::default(),
    )
        .expect("GLFW surface creation");

    let back_buffer = Framebuffer::back_buffer(surface.size());

    let pipeline_builder = surface.pipeline_builder();
    pipeline_builder.pipeline(&back_buffer, [0., 0., 0., 0.], |_pipeline, shd_gate| {
        let (program, warnings) =
            Program::<Semantics, (), ()>::from_strings(
                None, SIMPLE_VS, None, SIMPLE_FS,
            ).expect("program creation");

        for warning in warnings.iter() {
            eprintln!("Warning: {}", warning);
        }

        let tess1 = TessBuilder::new(&mut surface)
            .add_vertices(TRI_1_VERTICES)
            .set_mode(Mode::Triangle)
            .build().expect("Triangle 1 should tesselate");


        shd_gate.shade(&program, |rdr_gate, _| {
            rdr_gate.render(RenderState::default(), |tess_gate| {
                tess_gate.render(&mut surface, (&tess1).into());
            });
        });

        match bug_option {
            BugOption::Working => {}
            BugOption::Broken => {
                drop(tess1);
            }
        }

        let tess2 = TessBuilder::new(&mut surface)
            .add_vertices(TRI_2_VERTICES)
            .set_mode(Mode::Triangle)
            .build().expect("Triangle 2 should teselate");

        shd_gate.shade(&program, |rdr_gate, _| {
            rdr_gate.render(RenderState::default(), |tess_gate| {
                tess_gate.render(&mut surface, (&tess2).into());
            });
        });
    });

    surface.swap_buffers();

    'app: loop {
        // for all the events on the surface
        for event in surface.borrow_mut().poll_events() {
            match event {
                WindowEvent::Close | WindowEvent::Key(Key::Escape, _, Action::Release, _) => break 'app,

                WindowEvent::FramebufferSize(_width, _height) => {}

                _ => (),
            }
        }
    }
}```

**Comment by phaazon** [comment]
Hello!

Very good finding. I’m having a look right now! :)

**Comment by phaazon** [comment]
I found the bug. The problem is that dropping the tessellation implies freeing a resource for the next tessellation to create. The internal state tracker releases the `0` handle. Then the next tessellation uses that `0` handle. When it tries to `bind` it (that actually creates it), the state tracker already has `bound` the previous one (I do not unbind them!). Hence no binding hence no creation.

I’m patching it on the branch `fix/bind-tess` so that you can try. I’ll also try later in the night! Thanks again @austinjones!

**Comment by phaazon** [comment]
Whoops not fixed, I missed something.

**Comment by phaazon** [comment]
Ok so, I’m fixing that tomorrow. The fix will be part of the next milestone or if I find a way to make it a patch change (which should be the way), it should be released tomorrow on crates.io.

Thanks a lot for the feedback! I’ll have a look at your project! :)

**Comment by austinjones** [comment]
@phaazon Yay!  :tada: 

Thank you!

**Comment by phaazon** [comment]
![Screenshot 2019-09-11 at 14 13 04](https://user-images.githubusercontent.com/506592/64696278-4d669300-d49e-11e9-8b1e-667811054a5b.png)

Fixed! :)

**Comment by phaazon** [comment]
So unfortunately it’s not a patch change but a major one as I had to move some stuff around. It’s lightweight though, not a lot of code to change (basically, the back buffer must be obtained in another way and now requires a `GraphicsContext`).

I’ll ease things with `luminance-windowing` so that the back buffer is just trivial to get.

**Comment by phaazon** [comment]
@austinjones The fix is now live on [luminance-0.34](https://crates.io/crates/luminance). Feel free to have a look at the [changelog](https://github.com/phaazon/luminance-rs/blob/master/luminance/CHANGELOG.md#034) for information about how to migrate (should be easy for you).

**Comment by phaazon** [comment]
I updated the windowing crates. [You can get it even more easily now](https://github.com/phaazon/luminance-rs/blob/master/luminance/examples/01-hello-world.rs#L186):

```
let back_buffer = surface.back_buffer().unwrap(); // unwrap() is safe for now
```

🙄 🙂
